diff --git a/.github/workflows/wheels-dependencies.sh b/.github/workflows/wheels-dependencies.sh
index 8f47b3dea..74ab8b2e9 100755
--- a/.github/workflows/wheels-dependencies.sh
+++ b/.github/workflows/wheels-dependencies.sh
@@ -74,9 +74,23 @@ elif [[ "$(uname -s)" == "Darwin" ]]; then
     WORKDIR=$(pwd)/build/darwin
     BUILD_PREFIX=$(pwd)/build/deps/darwin
 else
-    # Build prefix will default to /usr/local
     PLAT="${CIBW_ARCHS:-$AUDITWHEEL_ARCH}"
     WORKDIR=$(pwd)/build
+
+    BUILD_PREFIX=~/.local
+    HOST_CONFIGURE_FLAGS="--host=aarch64-linux-android24 --disable-shared --enable-static"
+    HOST_CMAKE_FLAGS="-DCMAKE_SYSTEM_NAME=Android -DCMAKE_SYSTEM_PROCESSOR=aarch64 -DBUILD_SHARED_LIBS=NO -DENABLE_SHARED=NO"
+    HOST_MESON_FLAGS="--cross-file '$WORKDIR/meson-cross.txt' -Dprefer_static=true -Ddefault_library=static"
+    export PATH="$(dirname "$CC"):$PATH"
+    export LIBRARY_PATH="$BUILD_PREFIX/lib:${LIBRARY_PATH%:}"
+    export LD_LIBRARY_PATH="$LIBRARY_PATH"
+    export PKG_CONFIG_PATH="$BUILD_PREFIX/share/pkgconfig:$PKG_CONFIG_PATH"
+
+    git clone --depth 1 --branch v0.11.0 https://github.com/google/cpu_features.git build/cpu_features
+    sudo mkdir -p /sources/android
+    sudo ln -s "$(pwd)/build/cpu_features/ndk_compat" /sources/android/cpufeatures
+    export CFLAGS="$CFLAGS -I$(pwd)/build/cpu_features/include"
+
     MB_ML_LIBC=${AUDITWHEEL_POLICY::9}
     MB_ML_VER=${AUDITWHEEL_POLICY:9}
 fi
@@ -144,7 +158,7 @@ function build_zlib_ng {
     ORIGINAL_HOST_CONFIGURE_FLAGS=$HOST_CONFIGURE_FLAGS
     unset HOST_CONFIGURE_FLAGS
 
-    build_github zlib-ng/zlib-ng $ZLIB_NG_VERSION --installnamedir=$BUILD_PREFIX/lib --zlib-compat
+    CFLAGS="$CFLAGS -fPIC" build_github zlib-ng/zlib-ng $ZLIB_NG_VERSION --installnamedir=$BUILD_PREFIX/lib --zlib-compat
 
     HOST_CONFIGURE_FLAGS=$ORIGINAL_HOST_CONFIGURE_FLAGS
     touch zlib-stamp
@@ -175,6 +189,9 @@ function build_libavif {
     if [ -e libavif-stamp ]; then return; fi
 
     python3 -m pip install meson ninja
+    #ln -s "$CC" "$(dirname "$CC")/aarch64-linux-android21-clang"
+    echo megabruh
+    #readlink /usr/local/bin/aarch64-linux-android21-clang
 
     if ([[ "$PLAT" == "x86_64" ]] && [[ -z "$IOS_SDK" ]]) || [ -n "$SANITIZER" ]; then
         build_simple nasm 2.16.03 https://www.nasm.us/pub/nasm/releasebuilds/2.16.03
@@ -208,6 +225,14 @@ function build_libavif {
     fi
 
     local out_dir=$(fetch_unpack https://github.com/AOMediaCodec/libavif/archive/refs/tags/v$LIBAVIF_VERSION.tar.gz libavif-$LIBAVIF_VERSION.tar.gz)
+    sed -i -E "s|--cross-file=\"[^\"]+\"|--cross-file='$WORKDIR/meson-cross.txt'|" "$out_dir/ext/dav1d_android.sh"
+    (
+        head -1 "$out_dir/ext/dav1d_android.sh" \
+        && echo "cp '$WORKDIR/meson-cross.txt' '../../package/crossfiles/aarch64-android.meson'" \
+        && tail +2 "$out_dir/ext/dav1d_android.sh"
+    ) > patched.tmp
+    mv patched.tmp "$out_dir/ext/dav1d_android.sh"
+    cat "$out_dir/ext/dav1d_android.sh"
 
     # CONFIG_AV1_HIGHBITDEPTH=0 is a flag for libaom (included as a subproject
     # of libavif) that disables support for encoding high bit depth images.
@@ -228,14 +253,14 @@ function build_libavif {
             "${libavif_cmake_flags[@]}" \
             $HOST_CMAKE_FLAGS . )
 
-    if [[ -n "$IOS_SDK" ]]; then
+    if true; then
         # libavif's CMake configuration generates a meson cross file... but it
         # doesn't work for iOS cross-compilation. Copy in Pillow-generated
         # meson-cross config to replace the cmake-generated version.
-        cp $WORKDIR/meson-cross.txt $out_dir/crossfile-apple.meson
+        cp $WORKDIR/meson-cross.txt $out_dir/crossfile-android.meson
     fi
 
-    (cd $out_dir && make -j4 install)
+    (cd $out_dir && make -j4 install) || cat ~/test1
 
     touch libavif-stamp
 }
@@ -244,13 +269,13 @@ function build_zstd {
     if [ -e zstd-stamp ]; then return; fi
     local out_dir=$(fetch_unpack https://github.com/facebook/zstd/releases/download/v$ZSTD_VERSION/zstd-$ZSTD_VERSION.tar.gz)
     (cd $out_dir \
-        && make -j4 install)
+        && PREFIX="$BUILD_PREFIX" make -j4 install)
     touch zstd-stamp
 }
 
 function build {
     build_xz
-    if [ -z "$IS_ALPINE" ] && [ -z "$SANITIZER" ] && [ -z "$IS_MACOS" ]; then
+    if false; then
         yum remove -y zlib-devel
     fi
     if [[ -n "$IS_MACOS" ]]; then
@@ -260,7 +285,7 @@ function build {
     fi
 
     build_simple xcb-proto 1.17.0 https://xorg.freedesktop.org/archive/individual/proto
-    if [[ -n "$IS_MACOS" ]]; then
+    if true; then
         build_simple xorgproto 2025.1 https://www.x.org/pub/individual/proto
         build_simple libXau 1.0.12 https://www.x.org/pub/individual/lib
         build_simple libpthread-stubs 0.5 https://xcb.freedesktop.org/dist
@@ -317,7 +342,7 @@ function build {
 }
 
 function create_meson_cross_config {
-    cat << EOF > $WORKDIR/meson-cross.txt
+    cat << EOF > "$WORKDIR/meson-cross.txt"
 [binaries]
 pkg-config = '$BUILD_PREFIX/bin/pkg-config'
 cmake = '$(which cmake)'
@@ -332,9 +357,8 @@ c_link_args = '$CFLAGS -L$BUILD_PREFIX/lib'
 cpp_link_args = '$CFLAGS -L$BUILD_PREFIX/lib'
 
 [host_machine]
-system = 'darwin'
-subsystem = 'ios'
-kernel = 'xnu'
+system = 'linux'
+subsystem = 'linux'
 cpu_family = '$(uname -m)'
 cpu = '$(uname -m)'
 endian = 'little'
@@ -394,6 +418,8 @@ if [[ -n "$IS_MACOS" ]]; then
     fi
 fi
 
+create_meson_cross_config
+
 wrap_wheel_builder build
 
 # A safety catch for iOS. iOS can't use dynamic libraries, but clang will prefer
diff --git a/pyproject.toml b/pyproject.toml
index a2366e717..1b93640f5 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -3,6 +3,9 @@ build-backend = "backend"
 requires = [
   "pybind11",
   "setuptools>=77",
+  "pip",
+  "meson",
+  "ninja",
 ]
 backend-path = [
   "_custom_build",
@@ -97,7 +100,7 @@ dynamic.version = { attr = "PIL.__version__" }
 [tool.cibuildwheel]
 build-verbosity = 1
 config-settings = "raqm=enable raqm=vendor fribidi=vendor imagequant=disable"
-before-all = ".github/workflows/wheels-dependencies.sh"
+before-build = "cd build/pillow && .github/workflows/wheels-dependencies.sh"
 test-command = "cd {project} && .github/workflows/wheels-test.sh"
 test-requires = [
   "numpy",
