diff --git a/.github/workflows/wheels-dependencies.sh b/.github/workflows/wheels-dependencies.sh
index 8f47b3d..a477027 100755
--- a/.github/workflows/wheels-dependencies.sh
+++ b/.github/workflows/wheels-dependencies.sh
@@ -1,5 +1,12 @@
 #!/bin/bash
 
+set -e
+
+if [ -a ~/.pillow-dependencies-built ]; then
+    exit 0
+fi
+touch ~/.pillow-dependencies-built
+
 # Safety check - Pillow builds require that CIBW_ARCHS is set, and that it only
 # contains a single value (even though cibuildwheel allows multiple values in
 # CIBW_ARCHS). This check doesn't work on Linux because of how the CIBW_ARCHS
@@ -74,9 +81,24 @@ elif [[ "$(uname -s)" == "Darwin" ]]; then
     WORKDIR=$(pwd)/build/darwin
     BUILD_PREFIX=$(pwd)/build/deps/darwin
 else
-    # Build prefix will default to /usr/local
     PLAT="${CIBW_ARCHS:-$AUDITWHEEL_ARCH}"
     WORKDIR=$(pwd)/build
+
+    BUILD_PREFIX="$GITHUB_WORKSPACE/deps-prefix"
+    HOST_CONFIGURE_FLAGS="--host=aarch64-linux-android24 --disable-shared --enable-static"
+    HOST_CMAKE_FLAGS="-DCMAKE_SYSTEM_NAME=Android -DCMAKE_SYSTEM_PROCESSOR=aarch64 -DBUILD_SHARED_LIBS=NO -DENABLE_SHARED=NO"
+    HOST_MESON_FLAGS="--cross-file $WORKDIR/meson-cross.txt -Dprefer_static=true -Ddefault_library=static"
+    export CFLAGS="$CFLAGS -fPIC"
+    export PATH="$(dirname "$CC"):$PATH"
+    export LIBRARY_PATH="$BUILD_PREFIX/lib:${LIBRARY_PATH%:}"
+    export LD_LIBRARY_PATH="$LIBRARY_PATH"
+    export PKG_CONFIG_PATH="$BUILD_PREFIX/share/pkgconfig:$PKG_CONFIG_PATH"
+
+#    git clone --depth 1 --branch v0.11.0 https://github.com/google/cpu_features.git build/cpu_features
+#    sudo mkdir -p /sources/android
+#    sudo ln -s "$(pwd)/build/cpu_features/ndk_compat" /sources/android/cpufeatures
+#    export CFLAGS="$CFLAGS -I$(pwd)/build/cpu_features/include -DSTACK_LINE_READER_BUFFER_SIZE=1024"
+
     MB_ML_LIBC=${AUDITWHEEL_POLICY::9}
     MB_ML_VER=${AUDITWHEEL_POLICY:9}
 fi
@@ -244,13 +266,13 @@ function build_zstd {
     if [ -e zstd-stamp ]; then return; fi
     local out_dir=$(fetch_unpack https://github.com/facebook/zstd/releases/download/v$ZSTD_VERSION/zstd-$ZSTD_VERSION.tar.gz)
     (cd $out_dir \
-        && make -j4 install)
+        && PREFIX="$BUILD_PREFIX" make -j4 install)
     touch zstd-stamp
 }
 
 function build {
     build_xz
-    if [ -z "$IS_ALPINE" ] && [ -z "$SANITIZER" ] && [ -z "$IS_MACOS" ]; then
+    if false; then
         yum remove -y zlib-devel
     fi
     if [[ -n "$IS_MACOS" ]]; then
@@ -260,7 +282,7 @@ function build {
     fi
 
     build_simple xcb-proto 1.17.0 https://xorg.freedesktop.org/archive/individual/proto
-    if [[ -n "$IS_MACOS" ]]; then
+    if true; then
         build_simple xorgproto 2025.1 https://www.x.org/pub/individual/proto
         build_simple libXau 1.0.12 https://www.x.org/pub/individual/lib
         build_simple libpthread-stubs 0.5 https://xcb.freedesktop.org/dist
@@ -300,9 +322,11 @@ function build {
         https://storage.googleapis.com/downloads.webmproject.org/releases/webp tar.gz \
         --enable-libwebpmux --enable-libwebpdemux
 
+    rm -r "$BUILD_PREFIX/bin/"*
+
     build_brotli
 
-    if [[ -n "$IS_MACOS" ]]; then
+    if true; then
         # Custom freetype build
         build_simple freetype $FREETYPE_VERSION https://download.savannah.gnu.org/releases/freetype tar.gz --with-harfbuzz=no
     else
@@ -319,7 +343,6 @@ function build {
 function create_meson_cross_config {
     cat << EOF > $WORKDIR/meson-cross.txt
 [binaries]
-pkg-config = '$BUILD_PREFIX/bin/pkg-config'
 cmake = '$(which cmake)'
 c = '$CC'
 cpp = '$CXX'
@@ -332,9 +355,8 @@ c_link_args = '$CFLAGS -L$BUILD_PREFIX/lib'
 cpp_link_args = '$CFLAGS -L$BUILD_PREFIX/lib'
 
 [host_machine]
-system = 'darwin'
-subsystem = 'ios'
-kernel = 'xnu'
+system = 'linux'
+subsystem = 'linux'
 cpu_family = '$(uname -m)'
 cpu = '$(uname -m)'
 endian = 'little'
@@ -394,6 +416,8 @@ if [[ -n "$IS_MACOS" ]]; then
     fi
 fi
 
+create_meson_cross_config
+
 wrap_wheel_builder build
 
 # A safety catch for iOS. iOS can't use dynamic libraries, but clang will prefer
diff --git a/pyproject.toml b/pyproject.toml
index a2366e7..1b93640 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -3,6 +3,9 @@ build-backend = "backend"
 requires = [
   "pybind11",
   "setuptools>=77",
+  "pip",
+  "meson",
+  "ninja",
 ]
 backend-path = [
   "_custom_build",
@@ -97,7 +100,7 @@ dynamic.version = { attr = "PIL.__version__" }
 [tool.cibuildwheel]
 build-verbosity = 1
 config-settings = "raqm=enable raqm=vendor fribidi=vendor imagequant=disable"
-before-all = ".github/workflows/wheels-dependencies.sh"
+before-build = "cd build/pillow && .github/workflows/wheels-dependencies.sh"
 test-command = "cd {project} && .github/workflows/wheels-test.sh"
 test-requires = [
   "numpy",
diff --git a/setup.py b/setup.py
index f015a03..5ea9581 100644
--- a/setup.py
+++ b/setup.py
@@ -369,9 +369,7 @@ class pil_build_ext(build_ext):
         return True if value in configuration.get(option, []) else None
 
     def initialize_options(self) -> None:
-        self.disable_platform_guessing = self.check_configuration(
-            "platform-guessing", "disable"
-        )
+        self.disable_platform_guessing = True
         build_ext.initialize_options(self)
         for x in self.feature:
             setattr(self, f"disable_{x}", self.check_configuration(x, "disable"))
@@ -497,6 +495,12 @@ class pil_build_ext(build_ext):
         library_dirs: list[str] = []
         include_dirs: list[str] = []
 
+        ndk_root = os.path.join(
+            os.path.dirname(os.path.dirname(os.environ["CC"])), "sysroot", "usr"
+        )
+        _add_directory(library_dirs, os.path.join(ndk_root, "lib", "aarch64-linux-android", "24"))
+        _add_directory(include_dirs, os.path.join(ndk_root, "include"))
+
         pkg_config = None
         if _cmd_exists(os.environ.get("PKG_CONFIG", "pkg-config")):
             pkg_config = _pkg_config
@@ -705,6 +709,11 @@ class pil_build_ext(build_ext):
                 _add_directory(library_dirs, os.path.join(best_path, "lib"))
                 _add_directory(include_dirs, os.path.join(best_path, "include"))
 
+
+        deps_prefix = os.path.join(os.environ["GITHUB_WORKSPACE"], "deps-prefix")
+        _add_directory(library_dirs, os.path.join(deps_prefix, "lib"))
+        _add_directory(include_dirs, os.path.join(deps_prefix, "include"))
+
         #
         # insert new dirs *before* default libs, to avoid conflicts
         # between Python PYD stub libs and real libraries
@@ -905,7 +914,7 @@ class pil_build_ext(build_ext):
                 # so we have to guess; by default it is defined in all Windows builds.
                 # See #4237, #5243, #5359 for more information.
                 defs.append(("USE_WIN32_FILEIO", None))
-            elif sys.platform == "ios":
+            elif True:
                 # Ensure transitive dependencies are linked.
                 libs.append("lzma")
         if feature.get("jpeg"):
@@ -924,7 +933,7 @@ class pil_build_ext(build_ext):
             defs.append(("HAVE_LIBIMAGEQUANT", None))
         if feature.get("xcb"):
             libs.append(feature.get("xcb"))
-            if sys.platform == "ios":
+            if True:
                 # Ensure transitive dependencies are linked.
                 libs.append("Xau")
             defs.append(("HAVE_XCB", None))
@@ -959,9 +968,9 @@ class pil_build_ext(build_ext):
                     else:  # building FriBiDi shim from src/thirdparty
                         srcs.append("src/thirdparty/fribidi-shim/fribidi.c")
 
-            if sys.platform == "ios":
+            if True:
                 # Ensure transitive dependencies are linked.
-                libs.extend(["z", "bz2", "brotlicommon", "brotlidec", "png"])
+                libs.extend(["z", "brotlicommon", "brotlidec", "png"])
 
             self._update_extension("PIL._imagingft", libs, defs, srcs)
 
@@ -979,7 +988,7 @@ class pil_build_ext(build_ext):
         webp = feature.get("webp")
         if isinstance(webp, str):
             libs = [webp, webp + "mux", webp + "demux"]
-            if sys.platform == "ios":
+            if True:
                 # Ensure transitive dependencies are linked.
                 libs.append("sharpyuv")
             self._update_extension("PIL._webp", libs)
