diff --git a/.github/workflows/wheels-dependencies.sh b/.github/workflows/wheels-dependencies.sh
index 8f47b3d..6f73d23 100755
--- a/.github/workflows/wheels-dependencies.sh
+++ b/.github/workflows/wheels-dependencies.sh
@@ -74,9 +74,9 @@ elif [[ "$(uname -s)" == "Darwin" ]]; then
     WORKDIR=$(pwd)/build/darwin
     BUILD_PREFIX=$(pwd)/build/deps/darwin
 else
-    # Build prefix will default to /usr/local
     PLAT="${CIBW_ARCHS:-$AUDITWHEEL_ARCH}"
     WORKDIR=$(pwd)/build
+    BUILD_PREFIX=~/.local
     MB_ML_LIBC=${AUDITWHEEL_POLICY::9}
     MB_ML_VER=${AUDITWHEEL_POLICY:9}
 fi
@@ -250,9 +250,6 @@ function build_zstd {
 
 function build {
     build_xz
-    if [ -z "$IS_ALPINE" ] && [ -z "$SANITIZER" ] && [ -z "$IS_MACOS" ]; then
-        yum remove -y zlib-devel
-    fi
     if [[ -n "$IS_MACOS" ]]; then
         CFLAGS="$CFLAGS -headerpad_max_install_names" build_zlib_ng
     else
@@ -260,8 +257,8 @@ function build {
     fi
 
     build_simple xcb-proto 1.17.0 https://xorg.freedesktop.org/archive/individual/proto
-    if [[ -n "$IS_MACOS" ]]; then
-        build_simple xorgproto 2025.1 https://www.x.org/pub/individual/proto
+    if true; then
+        build_simple xorgproto 2025.1 https://www.x.org/pub/individual/proto || cat /home/runner/work/wheeldroid/wheeldroid/build/pillow/build/xorgproto-2025.1/config.log
         build_simple libXau 1.0.12 https://www.x.org/pub/individual/lib
         build_simple libpthread-stubs 0.5 https://xcb.freedesktop.org/dist
     else
diff --git a/pyproject.toml b/pyproject.toml
index a2366e7..15eb343 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -97,7 +97,7 @@ dynamic.version = { attr = "PIL.__version__" }
 [tool.cibuildwheel]
 build-verbosity = 1
 config-settings = "raqm=enable raqm=vendor fribidi=vendor imagequant=disable"
-before-all = ".github/workflows/wheels-dependencies.sh"
+before-all = "cd build/pillow && .github/workflows/wheels-dependencies.sh"
 test-command = "cd {project} && .github/workflows/wheels-test.sh"
 test-requires = [
   "numpy",
