diff --git a/.github/workflows/wheels-dependencies.sh b/.github/workflows/wheels-dependencies.sh
index b3a5782..15cbad9 100755
--- a/.github/workflows/wheels-dependencies.sh
+++ b/.github/workflows/wheels-dependencies.sh
@@ -74,9 +74,21 @@ elif [[ "$(uname -s)" == "Darwin" ]]; then
     WORKDIR=$(pwd)/build/darwin
     BUILD_PREFIX=$(pwd)/build/deps/darwin
 else
-    # Build prefix will default to /usr/local
     PLAT="${CIBW_ARCHS:-$AUDITWHEEL_ARCH}"
     WORKDIR=$(pwd)/build
+
+    BUILD_PREFIX=~/.local
+    HOST_CONFIGURE_FLAGS="--host=aarch64-linux-android24"
+    HOST_CMAKE_FLAGS="-DCMAKE_SYSTEM_NAME=Android -DCMAKE_SYSTEM_PROCESSOR=aarch64"
+    HOST_MESON_FLAGS="--cross-file '$WORKDIR/meson-cross.txt'"
+    export LIBRARY_PATH="$BUILD_PREFIX/lib:${LIBRARY_PATH%:}"
+    export LD_LIBRARY_PATH="$LIBRARY_PATH"
+    export PKG_CONFIG_PATH="$BUILD_PREFIX/share/pkgconfig:$PKG_CONFIG_PATH"
+
+    git clone --depth 1 --branch v0.11.0 https://github.com/google/cpu_features.git build/cpu_features
+    sudo mkdir -p /sources/android
+    sudo ln -s "$(pwd)/build/cpu_features/ndk_compat" /sources/android/cpufeatures
+
     MB_ML_LIBC=${AUDITWHEEL_POLICY::9}
     MB_ML_VER=${AUDITWHEEL_POLICY:9}
 fi
@@ -144,7 +156,7 @@ function build_zlib_ng {
     ORIGINAL_HOST_CONFIGURE_FLAGS=$HOST_CONFIGURE_FLAGS
     unset HOST_CONFIGURE_FLAGS
 
-    build_github zlib-ng/zlib-ng $ZLIB_NG_VERSION --installnamedir=$BUILD_PREFIX/lib --zlib-compat
+    CFLAGS="$CFLAGS -fPIC" build_github zlib-ng/zlib-ng $ZLIB_NG_VERSION --installnamedir=$BUILD_PREFIX/lib --zlib-compat
 
     HOST_CONFIGURE_FLAGS=$ORIGINAL_HOST_CONFIGURE_FLAGS
     touch zlib-stamp
@@ -208,6 +220,7 @@ function build_libavif {
     fi
 
     local out_dir=$(fetch_unpack https://github.com/AOMediaCodec/libavif/archive/refs/tags/v$LIBAVIF_VERSION.tar.gz libavif-$LIBAVIF_VERSION.tar.gz)
+    sed -i -E "s|(^meson setup .*)|\1 $HOST_MESON_FLAGS|" "$out_dir/ext/dav1d.cmd"
 
     # CONFIG_AV1_HIGHBITDEPTH=0 is a flag for libaom (included as a subproject
     # of libavif) that disables support for encoding high bit depth images.
@@ -228,11 +241,11 @@ function build_libavif {
             "${libavif_cmake_flags[@]}" \
             $HOST_CMAKE_FLAGS . )
 
-    if [[ -n "$IOS_SDK" ]]; then
+    if true; then
         # libavif's CMake configuration generates a meson cross file... but it
         # doesn't work for iOS cross-compilation. Copy in Pillow-generated
         # meson-cross config to replace the cmake-generated version.
-        cp $WORKDIR/meson-cross.txt $out_dir/crossfile-apple.meson
+        cp $WORKDIR/meson-cross.txt $out_dir/crossfile-android.meson
     fi
 
     (cd $out_dir && make -j4 install)
@@ -244,13 +257,13 @@ function build_zstd {
     if [ -e zstd-stamp ]; then return; fi
     local out_dir=$(fetch_unpack https://github.com/facebook/zstd/releases/download/v$ZSTD_VERSION/zstd-$ZSTD_VERSION.tar.gz)
     (cd $out_dir \
-        && make -j4 install)
+        && PREFIX="$BUILD_PREFIX" make -j4 install)
     touch zstd-stamp
 }
 
 function build {
     build_xz
-    if [ -z "$IS_ALPINE" ] && [ -z "$SANITIZER" ] && [ -z "$IS_MACOS" ]; then
+    if false; then
         yum remove -y zlib-devel
     fi
     if [[ -n "$IS_MACOS" ]]; then
@@ -260,7 +273,7 @@ function build {
     fi
 
     build_simple xcb-proto 1.17.0 https://xorg.freedesktop.org/archive/individual/proto
-    if [[ -n "$IS_MACOS" ]]; then
+    if true; then
         build_simple xorgproto 2025.1 https://www.x.org/pub/individual/proto
         build_simple libXau 1.0.12 https://www.x.org/pub/individual/lib
         build_simple libpthread-stubs 0.5 https://xcb.freedesktop.org/dist
@@ -396,6 +409,8 @@ if [[ -n "$IS_MACOS" ]]; then
     fi
 fi
 
+create_meson_cross_config
+
 wrap_wheel_builder build
 
 # A safety catch for iOS. iOS can't use dynamic libraries, but clang will prefer
diff --git a/pyproject.toml b/pyproject.toml
index 16d82d4..f52b6ab 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -3,6 +3,9 @@ build-backend = "backend"
 requires = [
   "pybind11",
   "setuptools>=77",
+  "pip",
+  "meson",
+  "ninja",
 ]
 backend-path = [
   "_custom_build",
@@ -97,7 +100,7 @@ dynamic.version = { attr = "PIL.__version__" }
 [tool.cibuildwheel]
 build-verbosity = 1
 config-settings = "raqm=enable raqm=vendor fribidi=vendor imagequant=disable"
-before-all = ".github/workflows/wheels-dependencies.sh"
+before-build = "cd build/pillow && .github/workflows/wheels-dependencies.sh"
 test-command = "cd {project} && .github/workflows/wheels-test.sh"
 test-requires = [
   "numpy",
