diff --git a/.github/workflows/wheels-dependencies.sh b/.github/workflows/wheels-dependencies.sh
index 8f47b3d..a415470 100755
--- a/.github/workflows/wheels-dependencies.sh
+++ b/.github/workflows/wheels-dependencies.sh
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -e
+
 # Safety check - Pillow builds require that CIBW_ARCHS is set, and that it only
 # contains a single value (even though cibuildwheel allows multiple values in
 # CIBW_ARCHS). This check doesn't work on Linux because of how the CIBW_ARCHS
@@ -74,9 +76,26 @@ elif [[ "$(uname -s)" == "Darwin" ]]; then
     WORKDIR=$(pwd)/build/darwin
     BUILD_PREFIX=$(pwd)/build/deps/darwin
 else
-    # Build prefix will default to /usr/local
     PLAT="${CIBW_ARCHS:-$AUDITWHEEL_ARCH}"
     WORKDIR=$(pwd)/build
+
+    BUILD_PREFIX="$GITHUB_WORKSPACE/deps-prefix"
+    HOST_CONFIGURE_FLAGS="--host=aarch64-linux-android24 --disable-shared --enable-static"
+    HOST_CMAKE_FLAGS="-DCMAKE_SYSTEM_NAME=Android -DCMAKE_SYSTEM_PROCESSOR=aarch64 -DBUILD_SHARED_LIBS=NO -DENABLE_SHARED=NO"
+    HOST_MESON_FLAGS="--cross-file $WORKDIR/meson-cross.txt -Dprefer_static=true -Ddefault_library=static"
+    export PATH="$(dirname "$CC"):$PATH"
+    export LIBRARY_PATH="$BUILD_PREFIX/lib:${LIBRARY_PATH%:}"
+    export LD_LIBRARY_PATH="$LIBRARY_PATH"
+    export PKG_CONFIG_PATH="$BUILD_PREFIX/share/pkgconfig:$PKG_CONFIG_PATH"
+
+    git clone --depth 1 --branch v0.11.0 https://github.com/google/cpu_features.git build/cpu_features
+    sudo mkdir -p /sources/android
+    sudo ln -s "$(pwd)/build/cpu_features/ndk_compat" /sources/android/cpufeatures
+    export CFLAGS="$CFLAGS -I$(pwd)/build/cpu_features/include -DSTACK_LINE_READER_BUFFER_SIZE=1024"
+
+    sudo dpkg --add-architecture arm64
+    sudo apt-get install glibc-headers:arm64
+
     MB_ML_LIBC=${AUDITWHEEL_POLICY::9}
     MB_ML_VER=${AUDITWHEEL_POLICY:9}
 fi
@@ -144,7 +163,7 @@ function build_zlib_ng {
     ORIGINAL_HOST_CONFIGURE_FLAGS=$HOST_CONFIGURE_FLAGS
     unset HOST_CONFIGURE_FLAGS
 
-    build_github zlib-ng/zlib-ng $ZLIB_NG_VERSION --installnamedir=$BUILD_PREFIX/lib --zlib-compat
+    CFLAGS="$CFLAGS -fPIC" build_github zlib-ng/zlib-ng $ZLIB_NG_VERSION --installnamedir=$BUILD_PREFIX/lib --zlib-compat
 
     HOST_CONFIGURE_FLAGS=$ORIGINAL_HOST_CONFIGURE_FLAGS
     touch zlib-stamp
@@ -244,13 +263,13 @@ function build_zstd {
     if [ -e zstd-stamp ]; then return; fi
     local out_dir=$(fetch_unpack https://github.com/facebook/zstd/releases/download/v$ZSTD_VERSION/zstd-$ZSTD_VERSION.tar.gz)
     (cd $out_dir \
-        && make -j4 install)
+        && PREFIX="$BUILD_PREFIX" make -j4 install)
     touch zstd-stamp
 }
 
 function build {
     build_xz
-    if [ -z "$IS_ALPINE" ] && [ -z "$SANITIZER" ] && [ -z "$IS_MACOS" ]; then
+    if false; then
         yum remove -y zlib-devel
     fi
     if [[ -n "$IS_MACOS" ]]; then
@@ -260,7 +279,7 @@ function build {
     fi
 
     build_simple xcb-proto 1.17.0 https://xorg.freedesktop.org/archive/individual/proto
-    if [[ -n "$IS_MACOS" ]]; then
+    if true; then
         build_simple xorgproto 2025.1 https://www.x.org/pub/individual/proto
         build_simple libXau 1.0.12 https://www.x.org/pub/individual/lib
         build_simple libpthread-stubs 0.5 https://xcb.freedesktop.org/dist
@@ -300,6 +319,8 @@ function build {
         https://storage.googleapis.com/downloads.webmproject.org/releases/webp tar.gz \
         --enable-libwebpmux --enable-libwebpdemux
 
+    rm -r "$BUILD_PREFIX/bin/"*
+
     build_brotli
 
     if [[ -n "$IS_MACOS" ]]; then
@@ -319,7 +340,6 @@ function build {
 function create_meson_cross_config {
     cat << EOF > $WORKDIR/meson-cross.txt
 [binaries]
-pkg-config = '$BUILD_PREFIX/bin/pkg-config'
 cmake = '$(which cmake)'
 c = '$CC'
 cpp = '$CXX'
@@ -332,9 +352,8 @@ c_link_args = '$CFLAGS -L$BUILD_PREFIX/lib'
 cpp_link_args = '$CFLAGS -L$BUILD_PREFIX/lib'
 
 [host_machine]
-system = 'darwin'
-subsystem = 'ios'
-kernel = 'xnu'
+system = 'linux'
+subsystem = 'linux'
 cpu_family = '$(uname -m)'
 cpu = '$(uname -m)'
 endian = 'little'
@@ -394,6 +413,10 @@ if [[ -n "$IS_MACOS" ]]; then
     fi
 fi
 
+create_meson_cross_config
+echo megabruh
+cat '/home/runner/work/wheeldroid/wheeldroid/build/pillow/build/meson-cross.txt'
+
 wrap_wheel_builder build
 
 # A safety catch for iOS. iOS can't use dynamic libraries, but clang will prefer
diff --git a/pyproject.toml b/pyproject.toml
index a2366e7..1b93640 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -3,6 +3,9 @@ build-backend = "backend"
 requires = [
   "pybind11",
   "setuptools>=77",
+  "pip",
+  "meson",
+  "ninja",
 ]
 backend-path = [
   "_custom_build",
@@ -97,7 +100,7 @@ dynamic.version = { attr = "PIL.__version__" }
 [tool.cibuildwheel]
 build-verbosity = 1
 config-settings = "raqm=enable raqm=vendor fribidi=vendor imagequant=disable"
-before-all = ".github/workflows/wheels-dependencies.sh"
+before-build = "cd build/pillow && .github/workflows/wheels-dependencies.sh"
 test-command = "cd {project} && .github/workflows/wheels-test.sh"
 test-requires = [
   "numpy",
diff --git a/setup.py b/setup.py
index f015a03..8d41a1e 100644
--- a/setup.py
+++ b/setup.py
@@ -709,6 +709,10 @@ class pil_build_ext(build_ext):
         # insert new dirs *before* default libs, to avoid conflicts
         # between Python PYD stub libs and real libraries
 
+        _dir = os.path.join(os.environ["GITHUB_WORKSPACE"], "deps-prefix")
+        _add_directory(library_dirs, os.path.join(_dir, "lib"))
+        _add_directory(include_dirs, os.path.join(_dir, "include"))
+
         self.compiler.library_dirs = library_dirs + self.compiler.library_dirs
         self.compiler.include_dirs = include_dirs + self.compiler.include_dirs
 
